### ISSUE TITLE
Feature: Observe File Changes

### ISSUE THREAD

**Issue by AjaniBilby** [8051Enthusiast/biodiff#11]
This little tool is an actual life saver, is there any chance you could add a feature where it watches the two input files, and then updates when they're changed?

Right now it appears to lock the files, which makes it a bit annoying when trying to debug output binaries from iterative testing

**Comment by 8051Enthusiast** [comment]
That's a good question - right now it actually `mmap`s the file so that changes would show up in unaligned view if you refresh the view, but in aligned mode, it buffers the old contents so i obviously didn't have that sort of thing in mind until now.

Maybe you have an idea how a refresh would work in aligned mode?
Because if there is an insertion, then the old alignment would be no longer ideal.
* I could of course ignore that and just update the displayed bytes, still using the offsets from the old alignment.
* I could add a key (say f5) that aligns the original bytes in the current buffer (say the currently view or selection if present) to the new file and then restarts alignment from there. This is would be a manual action since this is slow.
Do you have any other ideas?

**Comment by AjaniBilby** [comment]
> A `Mmap` may be backed by a file, or it can be anonymous map, backed by volatile memory.
> - https://docs.rs/memmap/latest/memmap/struct.Mmap.html

This isn't every explicit was to what `backed by a file` means, but if I have `biodiff` open on a file, I cannot write to it - and according to the OS the file is indeed locked by `biodiff`. So the question of how to detect updates is semi-irrelevant because updates cannot be performed to the file in use.

The only option I can think of that would allow you to still use `Mmap` would be to load the entire file into memory, then use that for it to be `backed` by - then you have a seperate file descriptor just waiting for file change, which then triggers a refresh.

But I presume the reason you used `Mmap` was because it doesn't require the entire file to be loaded in to memory?  
`Mmap` seems like a very heavy and restrictic abstraction (in terms of behaviour, not performance).

**Comment by 8051Enthusiast** [comment]
i didn't realize that it would lock the files on some OSes (Windows?). `biodiff` already falls back to reading the entire file into memory anyway if the mmap call fails. The memory usage from the alignment algorithms is fairly high anyway, so the pros of not using mmap (consistent view of files at one time, no locking) probably outweights the cons (slightly larger startup time on larger files).

**Comment by AjaniBilby** [comment]
Ohh no... It's possibly an upstream OS specific issue, that's the summary of my entire month xD

I am on Windows.
Here's the error which NodeJS spits out when trying to read a file open by `biodiff`, which is odd for a file error:
```
Error: UNKNOWN: unknown error, open './dump.wasm'
    at Object.openSync (node:fs:585:3)
    at Object.writeFileSync (node:fs:2229:35)
    at Object.<anonymous> (C:\Users\me\Documents\GitHub\salient\source\wasm\index.ts:51:4)
    at Module._compile (node:internal/modules/cjs/loader:1275:14)
    at Module.m._compile (C:\Users\me\AppData\Roaming\npm\node_modules\ts-node\src\index.ts:1618:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1329:10)
    at Object.require.extensions.<computed> [as .ts] (C:\Users\me\AppData\Roaming\npm\node_modules\ts-node\src\index.ts:1621:12)
    at Module.load (node:internal/modules/cjs/loader:1133:32)
    at Function.Module._load (node:internal/modules/cjs/loader:972:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:83:12) {
  errno: -4094,
  syscall: 'open',
  code: 'UNKNOWN',
  path: './dump.wasm'
}
```

**Comment by 8051Enthusiast** [comment]
after some thinking, i removed the mmap for now (the change is on the main branch), it's simply not worth it.
if someone does deem it necessary later, they can create an issue and i can probably add an option that enables mmap that is incompatible with updates, but it's probably only worth implementing if someone actually needs it.
in any case, the error you have should have disappeared.

i'll look into how to implement updates soon

**Comment by 8051Enthusiast** [comment]
I have decided that keeping track of updates in aligned mode is not really feasible - either the view ends up inconsistent or I would have to implement some heuristic that does not work half the time or easily causes out-of-memory problems.

What I have implemented though is the ability to reread the file in unaligned mode by pressing `F5`/`5`.
So if you are in the aligned view and want to see the updated file, you would have to switch to the unaligned view (F2), reread the file (F5) and then optionally move the views to your points of interests again if they moved with the file update, and then align again with F3.
